cairo/src/utils/signature.cairo

```diff
@@ -1,44 +1,352 @@
-from starkware.cairo.common.cairo_secp.ec import ec_add, ec_mul, ec_negate
 from starkware.cairo.common.uint256 import Uint256
 from starkware.cairo.common.alloc import alloc
 from src.utils.maths import unsigned_div_rem
 
 from src.interfaces.interfaces import ICairo1Helpers
+from src.utils.circuit_basic_field_ops import div_mod_p, neg_mod_p, is_opposite_mod_p, is_eq_mod_p
+from src.utils.circuit_utils import (
+    N_LIMBS,
+    hash_full_transcript_and_get_Z_3_LIMBS,
+    scalar_to_epns,
+    felt_to_UInt384,
+    run_modulo_circuit_basic,
+)
+from src.utils.uint256 import uint256_to_uint384
+
+from src.utils.ecdsa_circuit import (
+    get_full_ecip_2P_circuit,
+    get_ADD_EC_POINT_circuit,
+    get_DOUBLE_EC_POINT_circuit,
+)
+
+struct G1Point {
+    x: UInt384,
+    y: UInt384,
+}
+
+namespace secp256k1 {
+    const CURVE_ID = 2;
+    const P0 = 0xfffffffffffffffefffffc2f;
+    const P1 = 0xffffffffffffffffffffffff;
+    const P2 = 0xffffffffffffffff;
+    const P3 = 0x0;
+    const N0 = 0xaf48a03bbfd25e8cd0364141;
+    const N1 = 0xfffffffffffffffebaaedce6;
+    const N2 = 0xffffffffffffffff;
+    const N_LOW_128 = 0xbaaedce6af48a03bbfd25e8cd0364141;
+    const N_HIGH_128 = 0xfffffffffffffffffffffffffffffffe;
+    const N3 = 0x0;
+    const A0 = 0x0;
+    const A1 = 0x0;
+    const A2 = 0x0;
+    const A3 = 0x0;
+    const B0 = 0x7;
+    const B1 = 0x0;
+    const B2 = 0x0;
+    const B3 = 0x0;
+    const G0 = 0x3;
+    const G1 = 0x0;
+    const G2 = 0x0;
+    const G3 = 0x0;
+    const MIN_ONE_D0 = 0xfffffffffffffffefffffc2e;
+    const MIN_ONE_D1 = 0xffffffffffffffffffffffff;
+    const MIN_ONE_D2 = 0xffffffffffffffff;
+    const MIN_ONE_D3 = 0x0;
+}
+
+@known_ap_change
+func get_generator_point() -> (point: G1Point) {
+    // generator_point = (
+    //     0x79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798,
+    //     0x483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8
+    // ).
+    return (
+        point=G1Point(
+            x=UInt384(
+                0x2dce28d959f2815b16f81798, 0x55a06295ce870b07029bfcdb, 0x79be667ef9dcbbac, 0x0
+            ),
+            y=UInt384(
+                0xa68554199c47d08ffb10d4b8, 0x5da4fbfc0e1108a8fd17b448, 0x483ada7726a3c465, 0x0
+            ),
+        ),
+    );
+}
+
+@known_ap_change
+func sign_to_UInt384_mod_secp256k1(sign: felt) -> (res: UInt384) {
+    if (sign == -1) {
+        return (res=UInt384(secp256k1.MIN_ONE_D0, secp256k1.MIN_ONE_D1, secp256k1.MIN_ONE_D2, 0));
+    } else {
+        return (res=UInt384(1, 0, 0, 0));
+    }
+}
+
+// Takes a valid UInt384 value from ModuloBuiltin output, adds constraints to enforce it is canonically reduced mod p (ie: 0 <= a < p) and converts it to a Uint256.
+// Assumes 1 < p < 2^256 passed as valid Uint384.
+func uint384_to_uint256_mod_p{range_check_ptr}(a: UInt384, p: UInt384) -> (res: Uint256) {
+    assert a.d3 = 0;  // From assumption : "p < 2^256", canonical reduction of a is only possible if a.d3 is 0.
```

Comment:
I don't like this function as it mixes two different purposes without a real need. Let's just have a `uint384_to_uint256` that raises if it doesn't fit.

```
assert a.d3 = 0;
let (d2_high_64, d2_low_32) = divmod(a.d2, 2**64);
assert d2_high_64 = 0;
let (d1_high_64, d1_low_32) = divmod(a.d1, 2 ** 32);
// etc.
```

Reply:
and move to `uint256.cairo`